xapian-core/cluster/kmeans.cc

```diff
@@ -72,7 +79,7 @@ void
 KMeans::initialise_points(const MSet &source)
 {
     LOGCALL_VOID(API, "KMeans::initialise_points", source);
-    TermListGroup tlg(source);
+    tlg.construct_tlg(source);
```

Comment:
You can't really pass around an object directly in C++ - the choices are passing a pointer, passing a reference (which is essentially syntactic sugar for passing a pointer, except a `NULL` reference isn't valid), or passing by value (which actually means the object gets copied - that's cheap for simple types such as `int`, `double`, but passing a complex type by value is best avoided - certainly for containers like `std::map`, etc which could be arbitrarily large.  Since a `Stopper` can be an arbitrary user subclass, it could be arbitrarily expensive to copy.

Another reason you can't pass a `Stopper` by value is object slicing - see https://en.wikipedia.org/wiki/Object_slicing for a fuller explanation, but essentially if you pass a `FooStopper` to a `func(Stopper x)` then `x` will actually be a copy of just the `Stopper` base class.

The key consideration is who owns the pointed to `Stopper` object (and in particular who is responsible for making sure it gets destroyed at a suitable point, and that it only gets destroyed once).  Due to the `opt_intrusive_ptr` mechanism, a `Stopper` object might be reference counted (in which case the lifetime is automatically handled), or it isn't reference counted (in which case the caller is responsible).

The second case allows allocating it on the stack (which isn't possible with the non-"opt" `intrusive_ptr`) - e.g.:

```
{
    SimpleStopper stopper;
    stopper.add("no");
    QueryParser qp;
    QueryParser.set_stopper(&stopper);
    /// Use qp.
}
```

Here `stopper` lives until we leave the enclosing block, as does `qp`.

The way that `opt_intrusive_ptr` works is that `Stopper` has `opt_intrusive_base` as a base class, and that holds the reference count (this in the "intrusive" bit - some other approaches hold the reference count externally and can be used to reference count almost any class, but the trade-off is that the intrusive approach is simpler).  The reference count is `0` when the pointer isn't reference counted and one more than the number of references when it is.

So you can make `Xapian::Internal::opt_intrusive_ptr<Xapian::Stopper> stopper` a member of `KMeans` and then just pass `stopper.get()` to the `TermListGroup` constructor, which can take a `Stopper*`.

In this case, you don't actually even need to assign that `Stopper*` to another `opt_intrusive_ptr` in that constructor - we know the object must be valid while we're executing the `TermListGroup` constructor as either it's not reference counted, or the caller holds a reference (tricky cases involving a second thread aren't a concern, as we explicitly don't support using the same Xapian API object concurrently from multiple threads).  But if you were going to store such an object as a member, you can just put it into another `opt_intrusive_ptr`, which would either just hold it, or (if it's reference counted) increment the reference count and then decrement it again if that `opt_intrusive_ptr` is destroyed or assigned a new pointer.